src/main/java/com/atlassian/db/replica/internal/DecisionAwareReference.java

```diff
@@ -0,0 +1,44 @@
+package com.atlassian.db.replica.internal;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class DecisionAwareReference<T> {
+    private final LazyReference<T> lazyReference = new InnerLazyReference<>(this);
```

Comment:
I'm having trouble understanding why `LazyReference` wouldn't work

Reply:
Can you try to refactor? Maybe it's possible to use LazyReference, and I didn't figure out how.